### PR TITLE
Add support for creating rules in the Simplivity policy

### PR DESCRIPTION
**Pull request by svt-hchaudhari**
Signed-off-by: Hemant Chaudhari <hchaudhari@hpe.com>

### Description
Add support for creating rules in the Simplivity policy

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass(`$ tox`).
- [x] New functionality has been documented in the README if applicable.
  - [x] New functionality has been thoroughly documented in the examples (please include helpful comments).
  - [x] New endpoints supported are updated in the endpoints-support.md file.
- [x] Changes are documented in the CHANGELOG.

### CODE REVIEW

File: simplivity/resources/policies.py [hunk 1]
```diff
@@ -120,3 +121,48 @@ def delete(self, timeout=-1):
...
+                               If set to False, adds the new rules to existing set of rules
+            timeout: Time out for the request in seconds.
+            external_store_name: The name of the external_store
+
+        Returns:
+            self: Returns the same object.
+        """
+        resource_uri = "{}/{}/rules".format(URL, self.data["id"])
+        flags = {'replace_all_rules': replace_all_rules}
+        affected_object = self._client.do_post(resource_uri, rules, timeout, None, flags)[0]
```
Comment (sijeesh) [inline simplivity/resources/policies.py]: Should we validate the rules type here?
It should be an array of dictionaries, right?
Can we accept both dictionary(Single rule) and an array of dictionaries(multiple rules)

Comment (svt-hchaudhari) [reply simplivity/resources/policies.py]: Hi @sijeesh , @svt-kmcdaniel 
I do not think that we should be doing any validation for rule type. 
From the swagger the input required for rules is Array of rules elements as follows,
[
  {
    "destination_id": "<string value>",
    "frequency": 0,
    "retention": 0,
    "days": "<string value>",
    "start_time": "<string value>",
    "end_time": "<string value>",
    "application_consistent": false,
    "consistency_type": "NONE",
    "external_store_name": "<string value>"
  }
]

If we have only one elements then it is considered as one rule getting created and if we have multiple element then multiple rules are getting created. 
I am not sure if this is really named dictionaries that are being used as input here. 
Once we pass the array REST API does the validation and returns the appropriate response. 

This is how we have been doing so far for all APIs and we feel that it is best mechanism not to add any extra validation code for rules. 

In summary we feel since it is array of elements pass it onto REST API for further validation. 
Please let us know your views on the same.

File: simplivity/resources/policies.py [hunk 2]
```diff
@@ -120,3 +121,48 @@ def delete(self, timeout=-1):
...
+                                        Default: false
+                consistency_type: Set to DEFAULT for a snapshot backup
+                                  Set to VSS for a Microsoft Volume Shadow Copy Service backup
+                                  Set to NONE for crash-consistent backups
+                                  Default: NONE
+                destination_id: The unique identifier (UID) of the omnistack_cluster to store the backup
+                                Default: local omnistack_cluster
+                external_store_name: The name of the external_store
+                days: The days of the week (for example, Mon,Fri), or month (for example, 1,15) to take backups or
+                        "last" to specify the last day of each month
```
Comment (svt-kmcdaniel) [inline simplivity/resources/policies.py]: spacing doesn't match all other comments

File: simplivity/resources/policies.py [hunk 3]
```diff
@@ -120,3 +121,48 @@ def delete(self, timeout=-1):
...
+                        Default: All (that is, every day)
```
Comment (svt-kmcdaniel) [inline simplivity/resources/policies.py]: spacing doesn't match all other comments

File: simplivity/resources/policies.py [hunk 4]
```diff
@@ -120,3 +121,48 @@ def delete(self, timeout=-1):
...
+                            This time is local to the time zone of the Hypervisor Management System (HMS)
+                            Default: 00:00
+                end_time: The time to stop backing up, for example, 14:30
+                          This time is local to the time zone of the Hypervisor Management System (HMS)
+                          Default: 00:00
+
+            replace_all_rules: If set to True, replaces the existing rules with new rules
+                               If set to False, adds the new rules to existing set of rules
+            timeout: Time out for the request in seconds.
+            external_store_name: The name of the external_store
```
Comment (svt-kmcdaniel) [inline simplivity/resources/policies.py]: This isn't being handled in the method definition.   If it is a rule it needs to be with the other rule options

File: simplivity/resources/policies.py (hunk 1 above)
Comment (svt-kmcdaniel) [reply simplivity/resources/policies.py]: @svt-hchaudhari,

I believe what is being asked here is to have the method accept a single dict for a single rule, and an array of dictionaries for multiple rules.

Then inside the method convert it to an array if it is a single rule so that the total customer experience is better.

File: examples/policies.py [hunk 5]
```diff
@@ -83,7 +83,23 @@
 print("\n\ncreate policy")
 policy_name = "fixed_frequency_retention_policy"
 policy = policies.create(policy_name)
-print(policy, policy.data)
+print(f"{policy}")
+print(f"{pp.pformat(policy.data)} \n")
+rules = [
+    {
+        "start_time": "14:30",
```
Comment (svt-kmcdaniel) [inline examples/policies.py:90]: would many people have a start_time without a stated end_time

**Review by svt-kmcdaniel — CHANGES_REQUESTED** [review]

File: simplivity/resources/policies.py [hunk 6]
```diff
@@ -120,3 +121,48 @@ def delete(self, timeout=-1):
...
+
+        Returns:
+            self: Returns the same object.
+        """
+        resource_uri = "{}/{}/rules".format(URL, self.data["id"])
+        flags = {'replace_all_rules': replace_all_rules}
+        affected_object = self._client.do_post(resource_uri, rules, timeout, None, flags)[0]
+        policy_obj = self._policies.get_by_id(affected_object["object_id"])
+        self.data = policy_obj.data
```
Comment (svt-kmcdaniel) [inline simplivity/resources/policies.py]: should we have a policy._refresh instead?

Comment (svt-hchaudhari) [reply simplivity/resources/policies.py]: Added policy.refresh function to refresh the policy data.

File: examples/policies.py (hunk 5 above)
Comment (svt-hchaudhari) [reply examples/policies.py:90]: Added the end_time parameter along with start_time parameter

File: simplivity/resources/policies.py (hunk 2 above)
Comment (svt-hchaudhari) [reply simplivity/resources/policies.py]: Corrected the spacing for parameters

File: simplivity/resources/policies.py (hunk 3 above)
Comment (svt-hchaudhari) [reply simplivity/resources/policies.py]: Corrected the spacing for parameters

File: simplivity/resources/policies.py (hunk 4 above)
Comment (svt-hchaudhari) [reply simplivity/resources/policies.py]: This was duplicate and it was accidentally left in parameter. Removed this extra line.

File: simplivity/resources/policies.py (hunk 1 above)
Comment (svt-hchaudhari) [reply simplivity/resources/policies.py]: As mentioned above, corrected the parameter handling. 
If it is single rule accepted the parameters as dictionary and then converted the dictionary to array list.
Also added the check if it is not dictionary then input parameter is array of rules and passed it as it is to the POST call. 

@sijeesh @svt-kmcdaniel  It would be really nice if you could have look at this implementation and provide your views on the same.

**Comment by sijeesh** [comment]
@svt-hchaudhari , Sorry for the delayed review. 
Looks like you should rebase your branch and update your PR. I can see some changes from other PRs in this PR.
Fix the Travis build error also.

**Comment by svt-hchaudhari** [comment]
Closing this PRs due merge issue.
Opened new PR #33
